### PR TITLE
Take revenue

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #1115

### CODE REVIEW

File: runtime/karura/src/lib.rs [hunk 1]
```diff
@@ -1060,6 +1062,15 @@ impl OnUnbalanced<NegativeImbalance> for DealWithFees {
 		}
 	}
 }
+impl TakeRevenue for DealWithFees {
+	fn take_revenue(revenue: MultiAsset) {
+		if let MultiAsset::ConcreteFungible { amount, .. } = revenue {
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1067]: Not all concrete fungible revenue is in KSM, need to check the the location.

File: runtime/mandala/src/lib.rs [hunk 2]
```diff
@@ -1106,6 +1108,15 @@ impl OnUnbalanced<NegativeImbalance> for DealWithFees {
 		}
 	}
 }
+impl TakeRevenue for DealWithFees {
+	fn take_revenue(revenue: MultiAsset) {
+		if let MultiAsset::ConcreteFungible { amount, .. } = revenue {
```
Comment (shaunxw) [inline runtime/mandala/src/lib.rs:1113]: Same here.

File: runtime/mandala/src/lib.rs [hunk 3]
```diff
@@ -1106,6 +1108,15 @@ impl OnUnbalanced<NegativeImbalance> for DealWithFees {
...
+			// ensure KaruraTreasuryAccount have ed for DOT.
+			// Ignore the result.
+			let _ = Tokens::deposit(AUSD, &TreasuryAccount::get(), amount);
```
Comment (shaunxw) [inline runtime/mandala/src/lib.rs:1116]:
```suggestion
			let _ = Tokens::deposit(DOT, &TreasuryAccount::get(), amount);
```

File: runtime/karura/src/lib.rs [hunk 4]
```diff
@@ -1060,6 +1062,15 @@ impl OnUnbalanced<NegativeImbalance> for DealWithFees {
 		}
 	}
 }
+impl TakeRevenue for DealWithFees {
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1065]: I would prefer not to reuse the `DealWithFees`. We should make each struct type concern their own business. Declare a new struct before `XcmConfig` and name it `ToTreasury`.

**Review by shaunxw — CHANGES_REQUESTED** [review]

File: runtime/karura/src/lib.rs [hunk 5]
```diff
@@ -1340,6 +1342,19 @@ parameter_types! {
 
 pub type Barrier = (TakeWeightCredit, AllowTopLevelPaidExecutionFrom<All<MultiLocation>>);
 
+pub struct ToTreasury;
+impl TakeRevenue for ToTreasury {
+	fn take_revenue(revenue: MultiAsset) {
+		if let MultiAsset::ConcreteFungible { id, amount } = revenue {
+			if CurrencyIdConvert::convert(id).is_some() {
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1349]: `is_some` check only ensures `id` could be converted to a `CurrencyId` but it doesn't have to be KSM.

To handle it correctly, get the wrapped `currency_id` in result and deposit with it.

**Review by shaunxw — APPROVED** [review]
LGTM now. We need to do end to end test before merging it.

**Comment by zjb0807** [comment]
image


```
[
  {
    "phase": {
      "applyExtrinsic": 0
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 185277000,
          "class": "Mandatory",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x1e03",
      "data": [
        1
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x0003",
      "data": [
        "owxc9EpKLTtSjkRmzSytjmJtBgjVcMdNf5ZwCvR1hDjTGSE"
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x0b00",
      "data": [
        {
          "token": "KSM"
        },
        "owxc9EpKLTtSjkRmzSytjmJtBgjVcMdNf5ZwCvR1hDjTGSE",
        999952000000
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x0c02",
      "data": [
        {
          "token": "KSM"
        },
        "owxc9EpKLTtSjkRmzSytjmJtBgjVcMdNf5ZwCvR1hDjTGSE",
        999952000000
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x0c02",
      "data": [
        {
          "token": "KSM"
        },
        "qmmNufxeWaAVLMER2va1v4w2HbuU683c5gGtuxQG4fKTZSb",
        48000000
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x3502",
      "data": [
        "0xd0fc0b77005cdc143478c9d30527010bc62d51b8cb1b2c2d37643a3c0d5e14af",
        {
          "complete": 600000000
        }
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x1e04",
      "data": [
        600000000,
        "0x8ffc8978d9512ac029f7a829ebb1a360066c362c1a7e17bef6212178503b505e"
      ]
    },
    "topics": []
  },
  {
    "phase": {
      "applyExtrinsic": 1
    },
    "event": {
      "index": "0x0000",
      "data": [
        {
          "weight": 0,
          "class": "Mandatory",
          "paysFee": "Yes"
        }
      ]
    },
    "topics": []
  }
]
```

The treasury(`qmmNufxeWaAVLMER2va1v4w2HbuU683c5gGtuxQG4fKTZSb`) take revenue of XCM execution fee `48000000`.

**Comment by xlc** [comment]
How hard to write some tests for this?

**Comment by zjb0807** [comment]
> How hard to write some tests for this?

BTW,  learn about the xcm transaction process.

**Comment by shaunxw** [comment]
We only need to mock the execution of received XCM from Kusama in integration tests. You may refer to the commented out ones in integration tests. https://github.com/AcalaNetwork/Acala/blob/8c69db09747622d20abae1937adef234755c7cea/runtime/mandala/tests/integration_test.rs#L1654

**Pull request merged**